python/ray/data/tests/test_map.py

```diff
@@ -1551,6 +1551,100 @@ def test_random_sample_checks(ray_start_regular_shared):
         ray.data.range(1).random_sample(10)
 
 
+def get_expected_mask_indices(
+    num_blocks: int, num_rows_per_block: int, fraction: float, seed: int
+) -> np.ndarray:
+    """A helper function to verify random_sample with seed. This assumes
+    there is only one batch in each block.
+
+    Returns:
+        An array of indices of the sampled rows
+    """
+    # Seed sequence for each block: [batch_idx, task_idx, seed]
+    seeds = [[0, i, seed] for i in range(num_blocks)]
+
+    # Generate random numbers for each block
+    res = np.concatenate(
+        [np.random.default_rng(s).random(num_rows_per_block) for s in seeds]
+    )
+
+    # Sample the random numbers and get the indices
+    expected = np.where(res < fraction)[0]
+
+    return expected
+
+
+def test_random_sample_fixed_seed_0001(ray_start_regular_shared):
+    """Tests random_sample() with a fixed seed.
+
+    https://github.com/ray-project/ray/pull/51401
+
+    This test is to ensure that the random sampling is reproducible.
+    In the following example, we generate a deterministic seed sequence
+    for each block. Each block generates 10 ranndom numbers and we pick
+    10% of them. The indices from Ray Data should be the same as the
+    ones generated by numpy.
+    """
+    ds = ray.data.range(100, override_num_blocks=10).random_sample(
+        fraction=0.1, seed=1234
+    )
+
+    result = ds.to_pandas()["id"].to_numpy()
+
+    # Expected:
+    expected = get_expected_mask_indices(
+        num_blocks=10, num_rows_per_block=10, fraction=0.1, seed=1234
+    )
```

Comment:
we can probably just check against hard-coded expected results. 
otherwise, we'll need to also test get_expected_mask_indices is deterministic.